platform/config/config-osgi-admin-agent/src/main/java/org/codice/ddf/config/agent/osgi/ConfigAdminAgent.java

```diff
@@ -0,0 +1,629 @@
+  public void setFactories(Map<Class<? extends ConfigGroup>, String> factories) {
+    synchronized (lock) {
+      LOGGER.debug("ConfigAdminAgent:setFactories({})", factories);
+
+      // TODO: determine what mappings no longer exist so we can remove the corresponding config
+      // objs
+      // process all configured config instances we have to monitor
+      this.factories.clear();
+      this.factories.putAll(factories);
+      factories.forEach(this::findConfigMappingsFor);
+    }
+  }
+
+  @Override
+  public void serviceChanged(ServiceEvent event) {
+    synchronized (lock) {
+      final ServiceReference<?> ref = event.getServiceReference();
+      final int type = event.getType();
+      // DDF_Custom_Mime_Type_Resolver.b61903e6-59cb-4394-8060-03e803c56652
+      LOGGER.debug("ConfigAdminAgent::serviceChanged() - type = [{}], service = [{}]", type, ref);
+      if ((type == ServiceEvent.REGISTERED) || (type == ServiceEvent.MODIFIED)) {
+        // filter the pids that already exist in config admin as those would have alreay been
+        // processed
+        ConfigAdminAgent.servicePids(ref)
+            .filter(this::configurationDoesNotExist)
+            .map(mapper::getMapping)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .forEach(this::updateConfigObjectFor);
+        LOGGER.debug("done processing all PIDs for service: {}", ref);
+      } else {
+        LOGGER.debug("ignoring event [{}] for service [{}]", type, ref);
+      }
+    }
+  }
+
+  @Override
+  public void configurationEvent(ConfigurationEvent event) {
+    synchronized (lock) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(
+            "ConfigAdminAgent:configurationEvent(type={}, pid={}, factoryPid={})",
+            event.getType(),
+            event.getPid(),
+            event.getFactoryPid());
+      }
+      final String pid = event.getPid();
+
+      try {
+        switch (event.getType()) {
+          case ConfigurationEvent.CM_UPDATED:
+          case ConfigurationEvent.CM_LOCATION_CHANGED:
+            final ConfigurationAdmin cfgAdmin = getService(event.getReference());
+
+            getConfiguration(cfgAdmin, event.getPid()).ifPresent(this::updateConfiguration);
+            break;
+          case ConfigurationEvent.CM_DELETED:
+            cache.remove(pid);
+            return;
+          default:
+            return;
+        }
+      } catch (InvalidSyntaxException | IOException | ConfigMappingException e) { // ignore
+        LOGGER.warn("failed to process event for config object '{}': {}", pid, e.getMessage());
+        LOGGER.debug("config event failure: {}", e, e);
+      }
+    }
+  }
+
+  @Override
+  public void mappingChanged(ConfigMappingEvent event) {
+    synchronized (lock) {
+      LOGGER.debug("ConfigAdminAgent:mappingChanged({})", event);
+      if (event.getType() == Type.REMOVED) {
+        // just leave it alone - TODO: should we delete the corresponding config object???
```

Comment:
❗️ I am thinking yes. This would happen if all providers are suddenly removed because their services are deactivated. In such case, we no longer need the corresponding config object and it should be safe to delete them.

Reply:
Fixed